service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/aws/s3storageFolder/ControlledAwsS3StorageFolderResource.java

```diff
@@ -0,0 +1,218 @@
+import bio.terra.workspace.service.resource.model.WsmResourceFields;
+import bio.terra.workspace.service.resource.model.WsmResourceType;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.Optional;
+
+public class ControlledAwsS3StorageFolderResource extends ControlledResource {
+  private final String bucketName;
+  private final String prefix;
+
+  @JsonCreator
+  public ControlledAwsS3StorageFolderResource(
+      @JsonProperty("wsmResourceFields") WsmResourceFields resourceFields,
+      @JsonProperty("wsmControlledResourceFields")
+          WsmControlledResourceFields controlledResourceFields,
+      @JsonProperty("bucketName") String bucketName,
+      @JsonProperty("prefix") String prefix) {
+    super(resourceFields, controlledResourceFields);
+    this.bucketName = bucketName;
+    this.prefix = prefix;
+    validate();
+  }
+
+  private ControlledAwsS3StorageFolderResource(
+      ControlledResourceFields common, String bucketName, String prefix) {
+    super(common);
+    this.bucketName = bucketName;
+    this.prefix = prefix;
+    validate();
+  }
+
+  public ControlledAwsS3StorageFolderResource(
+      DbResource dbResource, String bucketName, String prefix) {
+    super(dbResource);
+    this.bucketName = bucketName;
+    this.prefix = prefix;
+    validate();
+  }
+
+  public static ControlledAwsS3StorageFolderResource.Builder builder() {
+    return new ControlledAwsS3StorageFolderResource.Builder();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  @SuppressWarnings("unchecked")
+  public <T> T castByEnum(WsmResourceType expectedType) {
+    if (getResourceType() != expectedType) {
+      throw new BadRequestException(String.format("Resource is not a %s", expectedType));
+    }
+    return (T) this;
+  }
+
+  // -- getters used in serialization --
+  @JsonProperty("wsmResourceFields")
+  public WsmResourceFields getWsmResourceFields() {
+    return super.getWsmResourceFields();
+  }
+
+  @JsonProperty("wsmControlledResourceFields")
+  public WsmControlledResourceFields getWsmControlledResourceFields() {
+    return super.getWsmControlledResourceFields();
+  }
+
+  public String getBucketName() {
+    return bucketName;
+  }
+
+  public String getPrefix() {
+    return prefix;
+  }
+
+  // -- getters not included in serialization --
+
+  @Override
+  @JsonIgnore
+  public WsmResourceType getResourceType() {
+    return WsmResourceType.CONTROLLED_AWS_S3_STORAGE_FOLDER;
+  }
+
+  @Override
+  @JsonIgnore
+  public WsmResourceFamily getResourceFamily() {
+    return WsmResourceFamily.AWS_S3_STORAGE_FOLDER;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  @JsonIgnore
+  public Optional<UniquenessCheckAttributes> getUniquenessCheckAttributes() {
+    return Optional.of(
+        new UniquenessCheckAttributes()
+            .uniquenessScope(UniquenessScope.GLOBAL)
+            .addParameter("prefix", prefix));
```

Comment:
I think this also needs a parameter for region if I understand the AWS infra properly. Since buckets are regional, it's valid in AWS to have both `my-us-bucket/foo/bar.txt` and `my-eu-bucket/foo/bar.txt` even though they have the same prefix

Reply:
It would be valid to have different folders with the same name in different buckets, but this would get more complicated if/when we have multiple Environments in a WSM instance that might have region overlaps.

I think the safe way to do this would be to include the bucket name rather than the region, but I still think that maybe we should side-step the issue via the global uniqueness check, and accept that we lose the ability to create `foo` in more than on workspace.  WDYT?  

_This assumes that I'm reading line 117 correctly (we can only have one folder named 'foo' in any Workspace managed by this WSM instance)._

(one thing that informs my desire to be conservative is the fact that, unlike GCS, [S3 has no "create if do not exist" option](https://repost.aws/questions/QUPOqd552USXuyHJVKC9yAHg/questions/QUPOqd552USXuyHJVKC9yAHg/can-we-do-write-object-if-not-exist-in-s3-similar-to-conditional-put-in-dynamo-db?) for atomic object creation so we are dependent on the database to avoid collisions)

Reply:
Unresolved, as there's more to discuss here.

Reply:
created https://verily.atlassian.net/browse/TERRA-519 so that it is not missed in other tickets. 
had marked it resolved only since it was tracked elsewhere and doesnt block this PR.

Reply:
this might be what we need 

prefix + account for the check. along with disallowing / in prefix
reason - the scope 'workspace' isnt going to work for storage since all folders across a wsm instance 

Bucket names are expected to be globally unique and not needed (thanks John)

Reply:
Leaving this as a global uniqueness check on prefix sgtm if we're not quite sure how we want to handle multiple environments, it's easier to start with the more restrictive validation and loosen than to go the other way

Reply:
handled in https://github.com/DataBiosphere/terra-workspace-manager/pull/1214